File: data structures and algorithms/cpp/08_007.cpp
#include <vector>
#include <deque>

class SlidingWindowExtremaProduct {
public:
    std::vector<int> findExtremaProducts(
            std::vector<int>& prices, int k) {
        int n = prices.size();
        std::vector<int> result;
        
        if (n == 0 || k == 0 || k > n) {
            return result;
        }
        
        std::deque<int> maxDeque;
        std::deque<int> minDeque;
        
        // Process the first window
        for (int i = 0; i < k; i++) {
            // Remove smaller elements from maxDeque
            while (!maxDeque.empty() && 
                   prices[i] >= prices[maxDeque.back()]) {
                maxDeque.pop_back();
            }
            maxDeque.push_back(i);
            
            // Remove larger elements from minDeque
            while (!minDeque.empty() && 
                   prices[i] <= prices[minDeque.back()]) {
                minDeque.pop_back();
            }
            minDeque.push_back(i);
        }
        
        // Calculate product for the first window
        result.push_back(prices[maxDeque.front()] * 
                        prices[minDeque.front()]);
        
        // Process the rest of the windows
        for (int i = k; i < n; i++) {
            // Remove elements outside the current window
            while (!maxDeque.empty() && 
                   maxDeque.front() <= i - k) {
                maxDeque.pop_front();
            }
            while (!minDeque.empty() && 
                   minDeque.front() <= i - k) {
                minDeque.pop_front();
            }
            
            // Remove smaller elements from maxDeque
            while (!maxDeque.empty() && 
                   prices[i] >= prices[maxDeque.back()]) {
                maxDeque.pop_back();
            }
            maxDeque.push_back(i);
            
            // Remove larger elements from minDeque
            while (!minDeque.empty() && 
                   prices[i] <= prices[minDeque.back()]) {
                minDeque.pop_back();
            }
            minDeque.push_back(i);
            
            // Calculate product for the current window
            result.push_back(prices[maxDeque.front()] * 
                            prices[minDeque.front()]);
        }
        
        return result;
    }
};

File: data structures and algorithms/cs/08_007.cs
using System.Collections.Generic;

public class SlidingWindowExtremaProduct
{
   public int[] FindExtremaProducts(
      int[] prices,
      int k)
   {
      int n = prices.Length;
      if (n == 0 || k == 0 || k > n)
      {
         return new int[0];
      }

      int[] result = new int[n - k + 1];
      var maxList = new List<int>();
      var minList = new List<int>();

      // Process first window
      for (int i = 0; i < k; i++)
      {
         // Remove smaller elements from maxList
         while (maxList.Count > 0 &&
            prices[i] >= prices[maxList[maxList.Count - 1]])
         {
            maxList.RemoveAt(maxList.Count - 1);
         }
         maxList.Add(i);

         // Remove larger elements from minList
         while (minList.Count > 0 &&
            prices[i] <= prices[minList[minList.Count - 1]])
         {
            minList.RemoveAt(minList.Count - 1);
         }
         minList.Add(i);
      }

      // Calculate first window product
      result[0] = prices[maxList[0]] *
         prices[minList[0]];

      // Process remaining windows
      for (int i = k; i < n; i++)
      {
         // Remove elements outside window
         while (maxList.Count > 0 &&
            maxList[0] <= i - k)
         {
            maxList.RemoveAt(0);
         }
         while (minList.Count > 0 &&
            minList[0] <= i - k)
         {
            minList.RemoveAt(0);
         }

         // Remove smaller elements from maxList
         while (maxList.Count > 0 &&
            prices[i] >= prices[maxList[maxList.Count - 1]])
         {
            maxList.RemoveAt(maxList.Count - 1);
         }
         maxList.Add(i);

         // Remove larger elements from minList
         while (minList.Count > 0 &&
            prices[i] <= prices[minList[minList.Count - 1]])
         {
            minList.RemoveAt(minList.Count - 1);
         }
         minList.Add(i);

         // Calculate current window product
         result[i - k + 1] = prices[maxList[0]] *
            prices[minList[0]];
      }

      return result;
   }
}